src/com/mendix/widget/StarRating/StarRating.ts

```diff
@@ -0,0 +1,129 @@
+        this.contextObject = object;
+        this.ownerReference = "System.owner";
+        this.updateRendering();
+        this.resetSubscriptions();
+
+        if (callback) {
+            callback();
+        }
+    }
+
+    updateRendering() {
+        render(createElement(RatingComponent, this.getProps()), this.domNode);
+    }
+
+    uninitialize(): boolean {
+        unmountComponentAtNode(this.domNode);
+
+        return true;
+    }
+
+    private getProps() {
+        // The contextObject entityType for "average" is different from that of "rating"
+        // InitialRate for rateType "average" comes from averageAttribute on campaignEntity
+        // InitialRate for rateType "rating" comes from rateAttribute on rateEntity
+        const initialRate = this.contextObject
+            ? this.rateType === "average"
+                ? this.contextObject.get(this.averageAttribute) as number
+                : this.contextObject.get(this.rateAttribute) as number
+            : 0 ;
+
+        return {
+            configurationError: this.hasValidConfiguration(),
+            initialRate,
+            handleOnChange: this.onChangeMicroflow && this.submitData.bind(this),
+            onRateMicroflow: this.onChangeMicroflow,
+            ownerGUID: this.contextObject.get(this.ownerReference) as string,
+            rateType: this.rateType,
+            readOnly: this.readOnly
+        };
+    }
+
+    private hasValidConfiguration(): string | undefined {
+        const errorMessage: string[] = [];
+        if (this.contextObject) {
+            if ((this.rateType === "average") && this.contextObject.getEntity() !== this.campaignEntity.split("/")[1]) {
+                errorMessage.push(" - For rate type 'average', the contextObject should be campaign entity");
+            }
+            if ((this.rateType === "rating") && this.contextObject.getEntity() !== this.rateEntity) {
+                errorMessage.push(` - For rate type 'Single', the contextObject be rate entity '${this.rateEntity}'`);
+            }
+            if (this.rateType === "rating" && !this.contextObject.isReference(this.ownerReference)) {
+                errorMessage.push(` - Context object has no User / Owner association to it`);
+            }
+            if (errorMessage.length) {
+                errorMessage.unshift("Configuration Error: ");
+            }
+        }
+        return errorMessage.length ? errorMessage.join("\n") : undefined;
+    }
+
+    private submitData(rate: number, microflowError: (error: Error) => void) {
+        if (this.contextObject) {
+            this.contextObject.set(this.rateAttribute, rate);
+            if (this.onChangeMicroflow) {
+                this.executeMicroflow(this.contextObject.getGuid(), this.onChangeMicroflow, microflowError);
+            }
+        }
+    }
+
+    private executeMicroflow(mendixGUID: string, microflow: string, error: (error: Error) => void) {
```

Comment:
GUID is an abbreviation for Globally Unique Identifier. previously we had agreed to have abbreviations eg XPath. Like this.